### ISSUE TITLE
Belos: Speeding up the GMRES time?

### ISSUE THREAD

**Issue by searhein** [trilinos/Trilinos#9979]
## Question
@trilinos/belos 

We (@mayrmt and I) have performed some scalability studies for Belos GMRES with a one-level Schwarz preconditioner from FROSch. In particular, we used the following Belos parameters:

```
<ParameterList name="Block GMRES">
  <Parameter name="Block Size" type="int" value="1"/>
  <Parameter name="Convergence Tolerance" type="double" value="1e-08"/>
  <Parameter name="Flexible Gmres" type="bool" value="false"/>
  <Parameter name="Maximum Iterations" type="int" value="500"/>
  <Parameter name="Output Frequency" type="int" value="1"/>
  <Parameter name="Show Maximum Residual Norm Only" type="bool" value="1"/>
</ParameterList>
```

For a test with 4096 MPI ranks (= processors), we have seen the following timings:

```
Belos: BlockGmresSolMgr total solve time: 37.1265 - 99.8982% [1] {min=37.1255, max=37.1316, std dev=0.000389575} <157, 3229, 687, 7, 0, 0, 0, 0, 0, 16>
|   Belos: ICGS[2]: Orthogonalization: 22.164 - 59.6985% [96] {min=21.95, max=22.6689, std dev=0.121371} <852, 635, 350, 547, 1661, 28, 5, 2, 0, 16>
|   |   Belos: ICGS[2]: Ortho (Norm): 1.82114 - 8.21668% [96] {min=0.0960775, max=5.10216, std dev=1.18545} <1152, 288, 24, 24, 1812, 180, 597, 3, 0, 16>
|   |   |   Belos::MVT::MvDot: 1.82062 - 99.9713% [96] {min=0.0956756, max=5.10185, std dev=1.18539} <1152, 288, 24, 24, 1812, 180, 597, 3, 0, 16>
|   |   |   Remainder: 0.000523361 - 0.0287381%
|   |   Belos::MVT::MvScale: 0.0185803 - 0.083831% [96] {min=0.0144991, max=0.0235985, std dev=0.00245491} <16, 1230, 638, 363, 29, 186, 330, 882, 384, 38>
|   |   Belos: ICGS[2]: Ortho (Inner Product): 12.2254 - 55.1589% [190] {min=10.3801, max=15.9917, std dev=1.19282} <1152, 232, 80, 1451, 438, 219, 501, 7, 0, 16>
|   |   |   Belos::MVT::MvTransMv: 12.224 - 99.9888% [190] {min=10.378, max=15.991, std dev=1.19283} <1152, 232, 80, 1451, 438, 219, 501, 7, 0, 16>
|   |   |   Remainder: 0.00136748 - 0.0111856%
|   |   Belos: ICGS[2]: Ortho (Update): 8.09554 - 36.5257% [190] {min=1.5531, max=11.7307, std dev=2.42635} <16, 0, 423, 177, 221, 1771, 24, 49, 263, 1152>
|   |   |   Belos::MVT::MvTimesMatAddMv: 8.09477 - 99.9904% [190] {min=1.55253, max=11.7299, std dev=2.42633} <16, 0, 424, 176, 221, 1771, 24, 49, 263, 1152>
|   |   |   Remainder: 0.000776691 - 0.00959406%
|   |   Remainder: 0.00331425 - 0.0149533%
|   Belos::MVT::MvAddMv: 0.00416453 - 0.0112171% [4] {min=0.00177065, max=0.0057178, std dev=0.00054133} <14, 2, 11, 147, 972, 758, 1023, 910, 212, 47>
|   Belos::MVT::MvNorm: 0.002798 - 0.00753638% [2] {min=0.00122577, max=0.0087995, std dev=0.000668423} <293, 1744, 1635, 396, 5, 0, 6, 1, 1, 15>
|   Belos: Operation Prec*x: 13.7517 - 37.04% [96] {min=13.2618, max=13.8367, std dev=0.0736212} <1, 17, 2, 108, 124, 0, 0, 55, 2269, 1520>
|   |   FROSch: OneLevelPreconditioner::apply (Level 1): 13.7492 - 99.9821% [96] {min=13.2595, max=13.8342, std dev=0.0736026} <1, 17, 2, 108, 124, 0, 0, 58, 2274, 1512>
|   |   Remainder: 0.00246834 - 0.0179494%
|   Belos: Operation Op*x: 1.11223 - 2.99577% [96] {min=0.937191, max=1.58369, std dev=0.126599} <1327, 627, 296, 352, 1096, 280, 61, 13, 32, 12>
|   Belos::MVT::MvTimesMatAddMv: 0.0733552 - 0.197582% [1] {min=0.0140146, max=0.0990519, std dev=0.0188862} <16, 0, 0, 384, 245, 1572, 223, 168, 52, 1436>
|   Remainder: 0.0183301 - 0.0493721%
```

Now, we were very surprised that the orthogonalization of GMRES takes significantly longer than the application of the preconditioner, which is supposed to be relatively expensive: we solve local problems of average size 2.2e+05 on each MPI rank.

Do you have any suggestions what we could do to speed up the timings for GMRES? Thank you very much in advance!

**Comment by cgcgcg** [comment]
Have you tried shorter restart length? I believe the default might be 100?

**Comment by dridzal** [comment]
Try adding `<Parameter name="Num Blocks" type="int" value="40"/>`.  This may increase the total number of iterations, but it should decrease the orthogonalization time significantly, assuming 100 is the default like @cgcgcg stated.

**Comment by srajama1** [comment]
@jennloe We should change the default to 40. This problem shows up once every couple of years.

**Comment by hkthorn** [comment]
It should be noted that the min and max timings for many components of the orthogonalization indicate either an imbalance in the distribution of the linear problem or issues with the communication on the machine.  If core linear algebra functions are this imbalanced in their performance, orthogonalization will always be an issue.  Changing the size of the Krylov subspace won't entirely negate that.

**Comment by jennloe** [comment]
@hkthorn Are you okay with changing the GMRES restart length default?

**Comment by jennloe** [comment]
@cgcgcg is correct that the load imbalance in `Belos::MVT::MvTimesMatAddMv` is concerning.  I presume this is all using Tpetra?

**Comment by searhein** [comment]
Thank you for the suggestions. Of course, I was aware of the fact that the orthogonalization in GMRES gets increasingly expensive with larger iteration counts. However, I often use GMRES and I had never seen any results where GMRES took even closely as much time as the preconditioner. 

Therefore, I just ran another test where I limited the iteration count to 50:
```
Belos: BlockGmresSolMgr total solve time: 14.1846 - 99.8101% [1] {min=14.1803, max=14.1856, std dev=0.00075531} <3, 19, 119, 95, 0, 0, 8, 636, 2667, 549>
|   Belos: ICGS[2]: Orthogonalization: 6.24524 - 44.0282% [51] {min=6.12318, max=6.51358, std dev=0.068801} <847, 617, 182, 691, 1460, 276, 1, 6, 0, 16>
|   |   Belos: ICGS[2]: Ortho (Norm): 0.553457 - 8.86206% [51] {min=0.0353885, max=1.43686, std dev=0.362917} <1256, 113, 47, 96, 1608, 120, 199, 641, 0, 16>
|   |   |   Belos::MVT::MvDot: 0.553184 - 99.9507% [51] {min=0.0352239, max=1.43669, std dev=0.362888} <1256, 113, 47, 96, 1608, 120, 199, 641, 0, 16>
|   |   |   Remainder: 0.000272606 - 0.0492551%
|   |   Belos::MVT::MvScale: 0.00999572 - 0.160053% [51] {min=0.00815053, max=0.0123997, std dev=0.00120783} <235, 642, 1412, 72, 113, 86, 205, 728, 502, 101>
|   |   Belos: ICGS[2]: Ortho (Inner Product): 3.45345 - 55.2974% [100] {min=2.92072, max=4.60827, std dev=0.351972} <1238, 130, 48, 1676, 125, 215, 641, 7, 0, 16>
|   |   |   Belos::MVT::MvTransMv: 3.45275 - 99.9796% [100] {min=2.91972, max=4.60789, std dev=0.351984} <1237, 131, 48, 1676, 125, 215, 641, 7, 0, 16>
|   |   |   Remainder: 0.000705135 - 0.0204182%
|   |   Belos: ICGS[2]: Ortho (Update): 2.22642 - 35.6499% [100] {min=0.456464, max=3.29808, std dev=0.73677} <16, 0, 648, 192, 121, 1607, 96, 48, 117, 1251>
|   |   |   Belos::MVT::MvTimesMatAddMv: 2.22602 - 99.9821% [100] {min=0.456168, max=3.29765, std dev=0.736752} <16, 0, 648, 192, 121, 1607, 96, 48, 117, 1251>
|   |   |   Remainder: 0.000399301 - 0.0179346%
|   |   Remainder: 0.00190806 - 0.0305522%
|   Belos::MVT::MvAddMv: 0.00247497 - 0.0174483% [2] {min=0.000876531, max=0.00377146, std dev=0.000580605} <12, 87, 425, 714, 412, 329, 952, 861, 219, 85>
|   Belos::MVT::MvNorm: 0.00154876 - 0.0109185% [1] {min=0.000386964, max=0.00312222, std dev=0.000585415} <171, 248, 1290, 493, 292, 387, 683, 440, 83, 9>
|   Belos: Operation Prec*x: 7.30024 - 51.4658% [51] {min=7.0409, max=7.34763, std dev=0.0388207} <1, 15, 11, 107, 118, 0, 0, 108, 2477, 1259>
|   |   FROSch: OneLevelPreconditioner::apply (Level 1): 7.29887 - 99.9813% [51] {min=7.03973, max=7.34613, std dev=0.0388183} <1, 15, 11, 107, 118, 0, 0, 109, 2465, 1270>
|   |   Remainder: 0.00136239 - 0.0186622%
|   Belos: Operation Op*x: 0.585906 - 4.13057% [50] {min=0.493295, max=0.828616, std dev=0.0712704} <1399, 670, 257, 168, 752, 752, 42, 9, 20, 27>
|   Belos::MVT::MvTimesMatAddMv: 0.0384747 - 0.271242% [1] {min=0.0102772, max=0.0522193, std dev=0.0102561} <16, 0, 51, 572, 283, 1517, 97, 72, 134, 1354>
|   Remainder: 0.0107638 - 0.0758833%
Remainder: 0.0144681 - 0.101804%
```

Now, the preconditioner and the GMRES times are in the same order of magnitude. Of course this is better, but still not quite what I was hoping for. The linear equation system should be more or less perfectly balanced since I am using a linear elasticity matrix created by Galeri. Of course, in this case, I could improve the Krylov time by using CG instead, but I would still have the same issue once I move to a non-spd problem. 

@jennloe @hkthorn The tests are using Tpetra through Xpetra. Does this indicate that there is some problem with the communication on the machine? Or could this be some problem with Tpetra?

**Comment by jennloe** [comment]
@searhein Would you mind sharing your test script with me?  I'd like to see if I can replicate these timings and dig into some pieces more closely.  I am in the process of setting up some Belos performance testing, so we could add something like this to the test queue to monitor it.

**Comment by dridzal** [comment]
@jennloe @searhein  The timer reports

`Belos: ICGS[2]: Ortho (Norm): 0.553457 - 8.86206% [51] {min=0.0353885, max=1.43686, std dev=0.362917}`

I believe this is timing the norm computation for a single vector.  Regardless of the algorithm, the 40x imbalance is very concerning.  Are the parallel maps set up correctly?  @searhein , to help figure this out, could you try CG?

**Comment by searhein** [comment]
@jennloe Not the whole code that I ran is pushed to Trilinos yet. But I think that the issue can also be reproduced with code that is already a test case in FROSch. I will double-check and post some instructions on how to reproduce this (in case it is not just an issue of the machine I am using). It would be great if you could take a closer look at this. Thank you!

@dridzal The matrix and map are coming from Galeri, and actually we already ran much larger computations (>100k MPI ranks) based on this test. Therefore, I hope that things are set up correctly. However, the larger runs have been performed some time ago (so some things might have changed in Trilinos since then), and we used CG at that time. I will run again with CG to see if the problem disappears.

**Comment by cgcgcg** [comment]
My recommendation would be to add a bunch of MPI barriers to the different timers. Load imbalance shows up at global synchronization points, but that might not be where it is created.

**Comment by searhein** [comment]
@cgcgcg I agree on that. In this case however, since the min and max timers are very similar for `Belos: ICGS[2]: Orthogonalization` as well as `Belos: Operation Prec*x`, my first thought was that the timers may actually be correct. However, you may be right, and I will add some barriers to be sure.

**Comment by searhein** [comment]
I am sorry for the very slow response. I am currently running a bit out of time because of several deadlines before the end of the year. However, I have now ran the same computation with CG:

```
Belos: BlockCGSolMgr total solve time: 16.5329 - 99.9157% [1] {min=16.5316, max=16.5382, std dev=0.000407916} <2, 2689, 1353, 36, 0, 0, 0, 0, 0, 16>
|   Belos::MVT::Assign: 0.00112804 - 0.00682301% [2] {min=0.000620049, max=0.00200886, std dev=8.59883e-05} <16, 10, 196, 3077, 772, 4, 0, 1, 7, 13>
|   Belos: Operation Prec*x: 14.6682 - 88.7212% [110] {min=14.1601, max=14.7512, std dev=0.083638} <3, 16, 42, 183, 8, 0, 0, 9, 2195, 1640>
|   |   FROSch: OneLevelPreconditioner::apply (Level 1): 14.6642 - 99.9728% [110] {min=14.158, max=14.7457, std dev=0.083623} <3, 16, 43, 184, 6, 0, 0, 8, 2101, 1735>
|   |   Remainder: 0.00398767 - 0.0271859%
|   Belos::MVT::MvDot: 0.357012 - 2.1594% [219] {min=0.230872, max=1.4983, std dev=0.10333} <3344, 500, 2, 231, 3, 0, 0, 0, 0, 16>
|   Belos::MVT::MvNorm: 0.0679419 - 0.410951% [111] {min=0.0570153, max=0.118089, std dev=0.00356699} <44, 3305, 724, 7, 0, 0, 0, 0, 0, 16>
|   Belos: Operation Op*x: 1.16535 - 7.04866% [109] {min=0.652394, max=1.2236, std dev=0.0342039} <16, 0, 0, 0, 0, 0, 0, 0, 2007, 2073>
|   Belos::MVT::MvAddMv: 0.266419 - 1.61145% [327] {min=0.10235, max=0.281046, std dev=0.0105393} <16, 0, 0, 0, 0, 0, 0, 0, 373, 3707>
|   Remainder: 0.00686657 - 0.0415329%
```

Now, I see the behavior that I expected: Almost all the time is spent on the preconditioner, and almost no time in the Krylov method. In total, the computation time went from 37.1 (see my first post) to 16.5 for roughly 100 iterations in both cases.

Of course, I can just go with CG for now, but nonetheless, I am wondering what the problem is with GMRES. Of course this could also be due to the machine, but with these timings GMRES is not competitive at all for now. In case, you have any suggestions how to improve this, I would be glad, because there might be another linear system coming which is not symmetric :-)

@jennloe I will share the tests when I find a bit more time.

**Comment by cgcgcg** [comment]
I ran the test case provided by @searhein on one of our systems, but cannot observe a similar behavior. @jennloe What kind of debug options can we add to or enable in Belos to help @searhein to pin down what is happening?

**Comment by searhein** [comment]
@cgcgcg @jennloe I would be happy to run another test on the same machine to figure out what the problem is. However, my computing time there is limited, such that I cannot do extensive testing. 

(Of course, I can use CG for now, but it would be great to figure out what the problem is)

**Comment by searhein** [comment]
Since my last post, we were able to circumvent this issue by using CG for SPD problems. However, we (in particular, @LeaSassmannshausen) now performed some new tests for Navier-Stokes, where we cannot use CG. For those new tests, we experienced relatively high iteration times for the orthogonalization again.

Are there any developments/updates on this?

Do you have any suggestions how to improve the times? If possible, we would not want to use shorter restart length because it is more difficult to discuss the iteration counts / numerical scalability of the preconditioners then.

Thank you in advance!

**Comment by jhux2** [comment]
@searhein Were you able to instrument Belos with additional barriers to minimize the chance that time is slipping between operations within the Krylov method?

**Comment by searhein** [comment]
@jhux2 We did not do that yet, but we can look into it. 

When we use CG for SPD problems, the time spent on the preconditioner is comparable, but the time spent in Belos is negligible. In GMRES, there is a significant amount of time additionally spent in the orthogonalization.

Moreover, we have some older results (more than 3 years old) for Navier-Stokes, where the preconditioner time was comparable to the times we see now, but the orthogonalization time in GMRES was also - more or less - negligible. This was also on an older machine, but I cannot imagine that, for newer architectures, the GMRES times are supposed to go up that much.

But we could try do some additional testing with barriers inside Belos to pin down where the time is actually spent. I will discuss this with @LeaSassmannshausen.

**Comment by searhein** [comment]
@Sbte @christiealappatt @maxfirmbach If I remember correctly from the EuroTUG meeting, you mentioned that you had similar experiences with Belos. Could you please share this. Thank you!

**Comment by christiealappatt** [comment]
@searhein I rechecked the issue I had with Orthogonalization. My problem was when running s-step GMRES on AMD CPUs. After a detailed analysis I found that the problem is that some BLAS routines were not performing optimally. I tried both with Intel MKL and AMD AOCL-BLIS but the performance were not optimal. I think the problem here is not related and the huge imbalance (difference in min. and max.) that you observed is something to look into.
